### PR TITLE
Add data source for secrets

### PR DESCRIPTION
**Pull request by Aflynn50**
## Description
**This PR only has very basic acceptance tests right now. It will not be possible to do an end to end test until #454 is in as without this it is not possible for applications to make use of the secret.**

To access a secret in juju we are already adding in a new resource that allows you to create a new secret. But there may often be cases where the user does not wish to store the raw secret values in the terraform plan, and add them instead some other way. In this use case, a data source is needed.

A data source is added that allows a secret specified by name in a particular model to be referenced in the terraform plan.

## Type of change

- Add new data source

## Environment

- Juju controller version: 
3.3 and above

## QA steps

Manual QA steps should be done to test this PR.
- `juju bootstrap lxd`
- `juju add-model default`
- `juju add-secret test1 token=123123
- Run the terraform plan below and observe no errors.
```tf
terraform {
  required_providers {
    juju = {
      version = "~> 0.11.0"
      source  = "juju/juju"
    }
  }
}

provider "juju" {}

data "juju_model" "default" {
  name = "default"
}

data "juju_secret" "datasource-secretName" {
  name = "test1"
  model = data.juju_model.default.name
}
```

### CODE REVIEW

**Review by cderici — APPROVED** [review]
LGTM, modulo some nits. QA went well, I see the id of the secret in the terraform state 👍

File: internal/provider/data_source_secrets.go [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+
+	// Read Terraform configuration state into the model
+	resp.Diagnostics.Append(req.Config.Get(ctx, &data)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	readSecretOutput, err := d.client.Secrets.ReadSecret(&juju.ReadSecretInput{
+		ModelName: data.Model.ValueString(),
+		Name:      data.Name.ValueStringPointer(),
```
[the remark this answers is not in the document]
Comment (Aflynn50) [reply internal/provider/data_source_secrets.go]: I've changed it to do that. I wasn't sure which of `IsNull` and `IsUnknown` to use when checking if the SecretID exists. I tested both and they both worked on their own but I've put both in the code just to be safe.

File: internal/provider/data_source_secrets_test.go [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+  value =  {
+    {{- range $key, $value := .SecretValue }}
+    "{{$key}}" = "{{$value}}"
+    {{- end }}
+  }
+}
+
+data "juju_secret" "secret_data_source" {
+  name = juju_secret.secret_resource.name
+  model = juju_model.{{.ModelName}}.name
```
Comment (hmlanigan) [inline internal/provider/data_source_secrets_test.go:64]: This test is a bit contrived.  But we need a good charm for this and full implementation of all 3 pieces for secrets to do that. 

Please add a ToDo.

File: internal/provider/data_source_secrets_test.go [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+	resource.Test(t, resource.TestCase{
+		PreCheck:                 func() { testAccPreCheck(t) },
+		ProtoV6ProviderFactories: frameworkProviderFactories,
+		Steps: []resource.TestStep{
+			{
+				Config: testAccDataSourceSecret(modelName, secretName, secretValue),
+				Check: resource.ComposeTestCheckFunc(
+					resource.TestCheckResourceAttr("data.juju_secret.secret_data_source", "model", modelName),
+					resource.TestCheckResourceAttr("data.juju_secret.secret_data_source", "name", secretName),
+					resource.TestCheckResourceAttrPair("data.juju_secret.secret_data_source", "secretID", "juju_secret.secret_resource", "secretID"),
```
Comment (hmlanigan) [inline internal/provider/data_source_secrets_test.go:37]: nice find.

File: internal/provider/data_source_secrets.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (hmlanigan) [reply internal/provider/data_source_secrets.go]: I sometimes settled on ValueString() == "". It was more concrete to me for strings.

File: docs/resources/secret.md [hunk 4]
```diff
@@ -0,0 +1,30 @@
+---
+# generated by https://github.com/hashicorp/terraform-plugin-docs
+page_title: "juju_secret Resource - terraform-provider-juju"
+subcategory: ""
+description: |-
+  A resource that represents a Juju secret.
+---
+
+# juju_secret (Resource)
```
Comment (hmlanigan) [inline docs/resources/secret.md:9]: I wonder if we're going to have a merge conflict somewhere, you shouldn't be picking up this file. The original commit needs an example.tf too.

File: internal/provider/data_source_secrets.go [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+	// context for the logging subsystem.
+	subCtx context.Context
+}
+
+// secretDataSourceModel is the juju data stored by terraform.
+// tfsdk must match secret data source schema attribute names.
+type secretDataSourceModel struct {
+	// Model to which the secret belongs.
+	Model types.String `tfsdk:"model"`
+	// Name of the secret to be updated or removed.
```
Comment (hmlanigan) [inline internal/provider/data_source_secrets.go:36]: s/to be updated or removed//.

Those words are not accurate.

**Review by hmlanigan — COMMENTED** [review]
Sorry Alastair, I gave another look and found more.

File: internal/provider/data_source_secrets.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (Aflynn50) [reply internal/provider/data_source_secrets.go]: Very reasonable, changed to that

File: internal/provider/data_source_secrets_test.go (hunk 2 above)
Comment (Aflynn50) [reply internal/provider/data_source_secrets_test.go:64]: Added, there is also one at the top of this PR description, am not planning on merging until its done

File: examples/data-sources/juju_secret/data-souce.tf [hunk 6]
```diff
@@ -0,0 +1,4 @@
+data "juju_secret" "this" {
```
[the remark this answers is not in the document]
Comment (Aflynn50) [reply examples/data-sources/juju_secret/data-souce.tf:1]: Very fair, I just copied the pattern from the other example.

File: internal/provider/data_source_secrets.go (hunk 5 above)
Comment (Aflynn50) [reply internal/provider/data_source_secrets.go:36]: Removed.

File: examples/data-sources/juju_secret/data-souce.tf [hunk 7]
```diff
@@ -0,0 +1,4 @@
...
+  model = "model-name"
+  name  = "secret-name"
+}
```
[the remark this answers is not in the document]
Comment (hmlanigan) [reply examples/data-sources/juju_secret/data-souce.tf]: That's okay, we'll land that before RC.

File: examples/data-sources/juju_secret/data-souce.tf (hunk 6 above)
[the remark this answers is not in the document]
Comment (hmlanigan) [reply examples/data-sources/juju_secret/data-souce.tf]: First step in breaking that pattern. Thx!

**Review by hmlanigan — APPROVED** [review]
LGTM. Thanks for all the last minute changes!

**Pull request merged**